flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -124,6 +124,9 @@
      */
     private final HashMap<String, InternalKvState<K, ?, ?>> keyValueStatesByName;
 
+    /** Unwrapped changelog states used for recovery (not wrapped into e.g. TTL). */
+    private final HashMap<String, ChangelogState> changelogStates;
```

Comment:
When will the `changelogStates` be cleared?

Reply:
I'm not sure it's safe to clear it as it can **potentially** be used later after recovery.
Otherwise, it can be cleared in `completeRestore`.

Reply:
Since you will apply all the states during recovery, I think it's safe to clear it after recovery completed.

Reply:
Done (9fa5fde).